bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/handler/ConnectedDriveBridgeHandler.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bmwconnecteddrive.internal.handler;
+
+import static org.openhab.binding.bmwconnecteddrive.internal.utils.Constants.ANONYMOUS;
+
+import java.util.Hashtable;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.bmwconnecteddrive.internal.ConnectedDriveConfiguration;
+import org.openhab.binding.bmwconnecteddrive.internal.discovery.VehicleDiscovery;
+import org.openhab.binding.bmwconnecteddrive.internal.dto.NetworkError;
+import org.openhab.binding.bmwconnecteddrive.internal.dto.discovery.Dealer;
+import org.openhab.binding.bmwconnecteddrive.internal.dto.discovery.VehiclesContainer;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.Constants;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.Converter;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.types.Command;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ConnectedDriveBridgeHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class ConnectedDriveBridgeHandler extends BaseBridgeHandler implements StringResponseCallback {
+    private final Logger logger = LoggerFactory.getLogger(ConnectedDriveBridgeHandler.class);
+    private HttpClientFactory httpClientFactory;
+    private BundleContext bundleContext;
+    private VehicleDiscovery discoveryService;
+    private ServiceRegistration<?> discoveryServiceRegstration;
+    private Optional<ConnectedDriveProxy> proxy = Optional.empty();
+    private Optional<ConnectedDriveConfiguration> configuration = Optional.empty();
+    private Optional<ScheduledFuture<?>> refreshJob = Optional.empty();
+    private Optional<String> troubleshootFingerprint = Optional.empty();
+
+    private static final String DISCOVERY_FINGERPRINT = "discovery-fingerprint";
+    private ChannelUID discoveryFingerprintChannel;
+
+    public ConnectedDriveBridgeHandler(Bridge bridge, HttpClientFactory hcf, BundleContext bc) {
+        super(bridge);
+        httpClientFactory = hcf;
+        bundleContext = bc;
+        discoveryService = new VehicleDiscovery(this);
+        discoveryServiceRegstration = bundleContext.registerService(DiscoveryService.class.getName(), discoveryService,
+                new Hashtable<>());
```

Comment:
I suggest that you change your discovery service into a ThingHandlerService so that you don't have to manage osgi registration of this class. It is a bit verbose to explain how to make the change but I've found another PR that shows exactly the changes that are required: https://github.com/openhab/openhab-addons/pull/6396/files

Reply:
I've done the changes and it works in the OH3 environment so BundleContext and service registration are completely removed.